cli/cmd/deploy/deploy.go

```diff
@@ -356,6 +336,106 @@ func verifyAccess(ctx context.Context, c *client.Client, githubURL string) (*adm
 	return ghRes, nil
 }
 
+func inferSourceConnectors(projectPath string) (map[connector][]string, error) {
+	sourcesPath := filepath.Join(projectPath, "sources")
+	sources, err := os.ReadDir(sourcesPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// connector to sources maps
+	csMap := make(map[connector][]string)
+	for _, s := range sources {
+		fileName := s.Name()
+		content, err := os.ReadFile(filepath.Join(sourcesPath, fileName))
+		if err != nil {
+			return nil, fmt.Errorf("error in reading file %v : %w", fileName, err)
+		}
+
+		src := source{}
+		if err := yaml.Unmarshal(content, &src); err != nil {
+			return nil, fmt.Errorf("error in unmarshalling yaml file %v : %w", fileName, err)
+		}
+
+		c := connector{Name: src.Type, Type: src.Type}
+		srcs, ok := csMap[c]
+		if !ok {
+			srcs = csMap[c]
+		}
+		srcName := strings.TrimSuffix(fileName, filepath.Ext(fileName))
+		srcs = append(srcs, srcName)
+		csMap[c] = srcs
+	}
+	return csMap, nil
+}
+
+type connector struct {
+	Name string
+	Type string
+}
+
+type source struct {
+	Type string `yaml:"type"`
+}
+
+func s3Prompt(c connector) map[string]string {
+	prop := make(map[string]string)
+	accessKeyID := fmt.Sprintf("connector.%s.aws_access_key_id", c.Name)
+	prop["aws_access_key_id"] = cmdutil.PasswordPrompt(accessKeyID)
+
+	secretAccessKey := fmt.Sprintf("connector.%s.aws_secret_access_key", c.Name)
+	prop["aws_secret_access_key"] = cmdutil.PasswordPrompt(secretAccessKey)
+
+	region := fmt.Sprintf("connector.%s.region", c.Name)
+	prop["region"] = cmdutil.InputPrompt(region, "us-east-1")
+	return prop
+}
+
+func gcsPrompt(c connector) map[string]string {
+	prop := make(map[string]string)
+	question := []*survey.Question{
+		{
+			Name: "credentials_json",
+			Prompt: &survey.Input{
+				Message: fmt.Sprintf("connector.%s.sa_key_json (enter path of file to load from)", c.Name),
+			},
+			Validate: func(any interface{}) error {
+				val := any.(string)
+				if val == "" {
+					// user can chhose to leave empty for public sources
+					return nil
+				}
+				_, err := os.Stat(val)
+				return err
+			},
+			Transform: func(any interface{}) interface{} {
+				val := any.(string)
+				if val == "" {
+					return ""
+				}
+				// ignoring error since PathError is already validated
+				content, _ := os.ReadFile(val)
+				return string(content)
+			},
+		},
+	}
+
+	creds := ""
+	if err := survey.Ask(question, &creds); err != nil {
+		exitWithFailure(fmt.Errorf("gcs credentials failed with error %w", err))
+	}
+
+	if creds != "" {
+		prop["gcs_credentials"] = creds
+	}
+	return prop
+}
+
+var connectorPrompts = map[string]func(connector) map[string]string{
+	"s3":  s3Prompt,
+	"gcs": gcsPrompt,
+}
```

Comment:
I'm reflecting on a more generic design. It would be great if we could have the connector spec in `runtime/connectors/<name>` define the necessary variables, as well as the introductory text with the docs link, and just auto-generate a prompt based on that here in the CLI.

If we move as much connector-specific logic as possible into `runtime/connectors`, it makes it easier to find places to update when e.g. changing/adding fields. (And also makes connectors more pluggable.)

Reply:
Yeah that makes sense. I wasn't sure if I should access runtime/connectors in CLI so I kept it here.
I will make the changes